relay/writer/file.go

```diff
@@ -0,0 +1,381 @@
+	stage writerStage
+
+	// underlying binlog writer,
+	// it will be created/started until needed.
+	out bw.Writer
+
+	filename sync2.AtomicString // current binlog filename
+	offset   sync2.AtomicInt64  // the file size may > 4GB
+}
+
+// NewFileWriter creates a FileWriter instances.
+func NewFileWriter(cfg *FileConfig) Writer {
+	w := &FileWriter{
+		cfg: cfg,
+	}
+	w.filename.Set(cfg.Filename) // set the startup filename
+	return w
+}
+
+// Start implements Writer.Start.
+func (w *FileWriter) Start() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.stage != stageNew {
+		return errors.Errorf("stage %s, expect %s, already started", w.stage, stageNew)
+	}
+	w.stage = stagePrepared
+
+	return nil
+}
+
+// Close implements Writer.Close.
+func (w *FileWriter) Close() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.stage == stageClosed {
+		return errors.New("already closed")
+	}
+	w.stage = stageClosed
+
+	var err error
+	if w.out != nil {
+		err = w.out.Close()
+	}
+
+	return errors.Trace(err)
+}
+
+// Recover implements Writer.Recover.
+func (w *FileWriter) Recover(p *parser.Parser) (*RecoverResult, error) {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	if w.stage != stagePrepared {
+		return nil, errors.Errorf("stage %s, expect %s, please start the writer first", w.stage, stagePrepared)
+	}
+
+	return w.doRecovering(p)
+}
+
+// WriteEvent implements Writer.WriteEvent.
+func (w *FileWriter) WriteEvent(ev *replication.BinlogEvent) (*Result, error) {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	if w.stage != stagePrepared {
+		return nil, errors.Errorf("stage %s, expect %s, please start the writer first", w.stage, stagePrepared)
+	}
+
+	switch ev.Event.(type) {
+	case *replication.FormatDescriptionEvent:
+		return w.handleFormatDescriptionEvent(ev)
+	case *replication.RotateEvent:
+		return w.handleRotateEvent(ev)
+	default:
+		return w.handleEventDefault(ev)
+	}
+}
+
+// Flush implements Writer.Flush.
+func (w *FileWriter) Flush() error {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	if w.stage != stagePrepared {
+		return errors.Errorf("stage %s, expect %s, please start the writer first", w.stage, stagePrepared)
+	}
+
+	if w.out != nil {
+		return w.out.Flush()
+	}
+	return nil
+}
+
+// handle FormatDescriptionEvent:
+//   1. close the previous binlog file
+//   2. open/create a new binlog file
+//   3. write the binlog file header if not exists
+//   4. write the FormatDescriptionEvent if not exists one
+//   5. record the offset of the opened/created binlog file
+func (w *FileWriter) handleFormatDescriptionEvent(ev *replication.BinlogEvent) (*Result, error) {
+	// close the previous binlog file
+	if w.out != nil {
+		log.Infof("[relay] closing previous underlying binlog writer with status %v", w.out.Status())
+		err := w.out.Close()
+		if err != nil {
+			return nil, errors.Annotate(err, "close previous underlying binlog writer")
+		}
+	}
+
+	// open/create a new binlog file
+	filename := filepath.Join(w.cfg.RelayDir, w.filename.Get())
```

Comment:
I think we should add a forcing file name correctness checking

Reply:
added checking, and some similar logic with _pkg/streamer/file.go_ can be refactored later.